pkg/fqdn/service/service.go

```diff
@@ -0,0 +1,446 @@
+	"github.com/cilium/cilium/pkg/logging"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+	"github.com/cilium/cilium/pkg/policy"
+	"github.com/cilium/cilium/pkg/time"
+
+	standalonednsproxy "github.com/cilium/cilium/api/v1/standalone-dns-proxy"
+)
+
+type updateOnDNSMsgFunc func(lookupTime time.Time, ep *endpoint.Endpoint, qname string, responseIPs []netip.Addr, TTL int, stat *dnsproxy.ProxyRequestContext) error
+
+type FQDNDataServer struct {
+	standalonednsproxy.UnimplementedFQDNDataServer
+
+	ctx             context.Context
+	closeServer     func()
+	endpointManager endpointmanager.EndpointManager
+
+	// streams is a map of the active streams and their cancel functions
+	// Streams are added when a client(standalone dns proxy) subscribes to DNS policies and removed when the client closes the connection
+	streams lock.Map[standalonednsproxy.FQDNData_StreamPolicyStateServer, context.CancelFunc]
+
+	// updateOnDNSMsg is a function to update the DNS message in the cilium agent on receiving the FQDN mapping
+	updateOnDNSMsg updateOnDNSMsgFunc
+
+	// dnsMappingResult is a map of the dns request id and bool value for success/failure
+	dnsMappingResult lock.Map[string, bool]
+
+	// snapshotMutex is a mutex to protect the current state of the DNS rules
+	snapshotMutex lock.Mutex
+
+	// currentSnapshot is the current state of the DNS rules
+	currentSnapshot map[identity.NumericIdentity]policy.SelectorPolicy
+
+	// identityToIpMutex is a mutex to protect the current state of the identity to Ip mapping
+	identityToIpMutex lock.Mutex
+
+	// currentIdentityToIp is a map of the identity to list of Ips
+	currentIdentityToIp map[identity.NumericIdentity][]net.IP
+}
+
+var (
+	log  = logging.DefaultLogger.WithField(logfields.LogSubsys, "fqdn/server")
+	kaep = keepalive.EnforcementPolicy{
+		PermitWithoutStream: true, // Allow pings even when there are no active streams
+	}
+	kasp = keepalive.ServerParameters{
+		Time:    5 * time.Second, // Ping the client if it is idle for 5 seconds to ensure the connection is still active
+		Timeout: 1 * time.Second, // Wait 1 second for the ping ack before assuming the connection is dead
+	}
+)
+
+// StreamPolicyState is a bidirectional streaming RPC to subscribe to DNS policies
+// SDP calls this method to subscribe to DNS policies
+// For each stream, we start a goroutine to receive the DNS policies ACKs
+// The flow of the method is as follows:
+// 1. Add the stream to the map( called by the client i.e SDP)
+// 2. Start a goroutine to receive the DNS policies ACKs for that particular client.
+// 3. Send the current state of the DNS rules to the client (We store the current state fo DNS rules during the endpoint regeneration see UpdatePolicyRulesLocked)
+// 4. Wait for the context to be done
+func (s *FQDNDataServer) StreamPolicyState(stream standalonednsproxy.FQDNData_StreamPolicyStateServer) error {
+	streamCtx, cancel := context.WithCancel(stream.Context())
+	s.streams.Store(stream, cancel)
+
+	go func() {
```

Comment:
If we pass around the cancel func instead of storing it on the server, we can avoid this. If you don't want to change that, I think this doesn't have to be a goroutine - it can just be a `defer` since the handler has the same lifecycle as the stream anyway.

Reply:
Agreed. Also since we removed the context listening of the server I removed the listener from the bottom as well as it was redundant.